qutip/core/coefficient.py

```diff
@@ -498,8 +499,15 @@ def compile_code(code, file_name, parsed, c_opt):
                 include_dirs=[np.get_include()],
                 language='c++'
             )
+            compile_time_env = {
+                "QUTIP_IDXINT_64": idxint_dtype is np.int64,
+            }
             with _ignore_import_warning_for_pyximporter():
-                ext_modules = cythonize(coeff_file, force=c_opt['recompile'])
+                ext_modules = cythonize(
+                    coeff_file,
+                    compile_time_env=compile_time_env,
+                    force=c_opt['recompile'],
+                )
```

Comment:
The `int32` vs `int64` is for the data in `CSR`, it does not affect the coefficient. If needed, the coefficient should cimport `base.idxint` instead of using environment variable.